### PR TITLE
Add App-scoped API For Firebase Project management

### PR DESCRIPTION
**Pull request by xiangkangjw**
App-scoped actions for Firebase apps within Firebase projects. Uses Firebase Management REST APIs (https://firebase.google.com/docs/projects/api/reference/rest/)

### CODE REVIEW

**Comment by schmidt-sebastian** [comment]
This PR only contains previously reviewed changes.

**Comment by weixifan** [comment]
I fixed some iOS vs Android mistakes in the Javadoc for various methods in `FirebaseProjectManagement`. I will assign a reviewer in case there are comments about that change.

**Comment by weixifan** [comment]
I also just adjusted the error codes list. Both Android certificates and Apps are handled by the same error codes map, so I changed the term "App" to simply "resource." Please take a look.

**Comment by schmidt-sebastian** [comment]
Still LGTM

**Comment by weixifan** [comment]
Unfortunately I discovered more things to fix. :(

File: src/test/java/com/google/firebase/projectmanagement/FirebaseProjectManagementServiceImplTest.java [hunk 1]
```diff
@@ -0,0 +1,911 @@
...
+    GET,
+    POST,
+    DELETE
+  }
+
+  /**
+   * Can be used to intercept multiple HTTP requests and responses made by the SDK during tests.
+   */
+  private class MultiRequestTestResponseInterceptor implements HttpResponseInterceptor {
+    private List<HttpResponse> responsesList = new ArrayList<HttpResponse>();
```
Comment (hiranya911) [inline src/test/java/com/google/firebase/projectmanagement/FirebaseProjectManagementServiceImplTest.java]: private final

Comment (hiranya911) [inline src/test/java/com/google/firebase/projectmanagement/FirebaseProjectManagementServiceImplTest.java]: Use diamond syntax (we are on Java 7 where it's allowed):

```
new ArrayList<>();
```

File: src/test/java/com/google/firebase/projectmanagement/FirebaseProjectManagementServiceImplTest.java [hunk 2]
```diff
@@ -0,0 +1,911 @@
...
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+
```
Comment (hiranya911) [inline src/test/java/com/google/firebase/projectmanagement/FirebaseProjectManagementServiceImplTest.java]: There's a ton of code duplication in this file. Can we do something about it? One simple solution is to combine some tests into a single method as follows (you might have to change how mock responses are set up):

```
  @Test
  public void getIosApp() throws Exception {
    String expectedUrl = String.format(
        "%s/v1beta1/projects/-/iosApps/%s", FIREBASE_PROJECT_MANAGEMENT_URL, IOS_APP_ID);
    firstRpcResponse.setContent(String.format(GET_IOS_RESPONSE, DISPLAY_NAME));
    serviceImpl = initServiceImpl(firstRpcResponse, interceptor);
    IosAppMetadata iosAppMetadata = serviceImpl.getIosApp(IOS_APP_ID);
     checkRequestHeader(expectedUrl, HttpMethod.GET);
     assertEquals(iosAppMetadata, IOS_APP_METADATA);
  
     iosAppMetadata = serviceImpl.getIosAppAsync(IOS_APP_ID).get();
     checkRequestHeader(expectedUrl, HttpMethod.GET);
     assertEquals(iosAppMetadata, IOS_APP_METADATA);
  }
```

File: src/main/java/com/google/firebase/projectmanagement/FirebaseProjectManagementServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,909 @@
...
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+class FirebaseProjectManagementServiceImpl implements AndroidAppService, IosAppService {
+  @VisibleForTesting static final String FIREBASE_PROJECT_MANAGEMENT_URL =
```
Comment (hiranya911) [inline src/main/java/com/google/firebase/projectmanagement/FirebaseProjectManagementServiceImpl.java]: Add newline

File: src/main/java/com/google/firebase/projectmanagement/FirebaseProjectManagementServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,909 @@
...
+      new CreateAndroidAppFromAppIdFunction();
+  private final CreateIosAppFromAppIdFunction createIosAppFromAppIdFunction =
+      new CreateIosAppFromAppIdFunction();
+
+  private HttpResponseInterceptor interceptor;
+
+  FirebaseProjectManagementServiceImpl(FirebaseApp app) {
+    this.app = app;
+
+    JsonFactory jsonFactory = app.getOptions().getJsonFactory();
```
Comment (hiranya911) [inline src/main/java/com/google/firebase/projectmanagement/FirebaseProjectManagementServiceImpl.java]: Assign directly to `this.jsonFactory`

File: src/main/java/com/google/firebase/projectmanagement/FirebaseProjectManagementServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,909 @@
...
+    this.jsonFactory = jsonFactory;
+
+    HttpTransport httpTransport = app.getOptions().getHttpTransport();
+    this.requestFactory =
+        httpTransport.createRequestFactory(new FirebaseRequestInitializer(app));
+
+    this.scheduledExecutor = new FirebaseScheduledExecutor(
```
Comment (hiranya911) [inline src/main/java/com/google/firebase/projectmanagement/FirebaseProjectManagementServiceImpl.java]: Please remove this. No service should initialize its own thread pools in the SDK.

File: src/main/java/com/google/firebase/projectmanagement/FirebaseProjectManagementServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,909 @@
...
+  private CallableOperation<AndroidAppMetadata, FirebaseProjectManagementException> getAndroidAppOp(
+      final String appId) {
+    return new CallableOperation<AndroidAppMetadata, FirebaseProjectManagementException>() {
+      @Override
+      protected AndroidAppMetadata execute() throws FirebaseProjectManagementException {
+        String url = String.format(
+            "%s/v1beta1/projects/-/androidApps/%s", FIREBASE_PROJECT_MANAGEMENT_URL, appId);
+        AndroidAppResponse parsedResponse = new AndroidAppResponse();
+        makeGetRequest(url, parsedResponse, appId, "App ID");
+        return new AndroidAppMetadata(
```
Comment (hiranya911) [inline src/main/java/com/google/firebase/projectmanagement/FirebaseProjectManagementServiceImpl.java]: I think AndroidAppMetadata should just provide a constructor that accepts an AndroidAppResponse as the only argument.

File: src/main/java/com/google/firebase/projectmanagement/FirebaseProjectManagementServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,909 @@
...
+  private CallableOperation<IosAppMetadata, FirebaseProjectManagementException> getIosAppOp(
+      final String appId) {
+    return new CallableOperation<IosAppMetadata, FirebaseProjectManagementException>() {
+      @Override
+      protected IosAppMetadata execute() throws FirebaseProjectManagementException {
+        String url = String.format(
+            "%s/v1beta1/projects/-/iosApps/%s", FIREBASE_PROJECT_MANAGEMENT_URL, appId);
+        IosAppResponse parsedResponse = new IosAppResponse();
+        makeGetRequest(url, parsedResponse, appId, "App ID");
+        return new IosAppMetadata(
```
Comment (hiranya911) [inline src/main/java/com/google/firebase/projectmanagement/FirebaseProjectManagementServiceImpl.java]: Same here. Provide a constructor that takes IosAppResponse as the only argument.

File: src/main/java/com/google/firebase/projectmanagement/FirebaseProjectManagementServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,909 @@
...
+        String.format(
+            "%s \"%s\": %s", requestIdentifierDescription, requestIdentifier, messageBody),
+        cause);
+  }
+
+  /* Helper types. */
+
+  private interface CreateAppFromAppIdFunction<T> extends ApiFunction<String, T> {}
+
+  private class CreateAndroidAppFromAppIdFunction
```
Comment (hiranya911) [inline src/main/java/com/google/firebase/projectmanagement/FirebaseProjectManagementServiceImpl.java]: Make these static classes. Pass the service impl instance as a constructor arg.

File: src/main/java/com/google/firebase/projectmanagement/FirebaseProjectManagementServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,909 @@
...
+      @Nullable Exception cause) {
+    return new FirebaseProjectManagementException(
+        String.format(
+            "%s \"%s\": %s", requestIdentifierDescription, requestIdentifier, messageBody),
+        cause);
+  }
+
+  /* Helper types. */
+
+  private interface CreateAppFromAppIdFunction<T> extends ApiFunction<String, T> {}
```
Comment (hiranya911) [inline src/main/java/com/google/firebase/projectmanagement/FirebaseProjectManagementServiceImpl.java]: Lets remove this. Implementing types can directly implement ApiFunction interface.

File: src/main/java/com/google/firebase/projectmanagement/FirebaseProjectManagementServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,909 @@
...
+    }
+
+    /**
+     * Returns an {@link ApiFuture} that will eventually contain the App ID of the new created App,
+     * or an exception if an error occurred during polling.
+     */
+    @Override
+    public ApiFuture<String> apply(String operationName) throws FirebaseProjectManagementException {
+      SettableApiFuture<String> settableFuture = SettableApiFuture.<String>create();
+      scheduledExecutor.schedule(
```
Comment (hiranya911) [inline src/main/java/com/google/firebase/projectmanagement/FirebaseProjectManagementServiceImpl.java]: Expose the FirebaseApp.schedule() method via a trampoline, and use it here.

File: src/main/java/com/google/firebase/projectmanagement/FirebaseProjectManagementServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,909 @@
...
+      this.operationName = operationName;
+      this.projectId = projectId;
+      this.settableFuture = settableFuture;
+    }
+
+    @Override
+    public void run() {
+      String url = String.format("%s/v1/%s", FIREBASE_PROJECT_MANAGEMENT_URL, operationName);
+      OperationResponse operationResponseInstance;
+      operationResponseInstance = new OperationResponse();
```
Comment (hiranya911) [inline src/main/java/com/google/firebase/projectmanagement/FirebaseProjectManagementServiceImpl.java]: Combine with the previous line.

File: src/main/java/com/google/firebase/projectmanagement/FirebaseProjectManagementServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,909 @@
...
+        "Bundle ID",
+        "Unable to create iOS App");
+  }
+
+  @Override
+  public ApiFuture<IosApp> createIosAppAsync(
+      String projectId, String bundleId, String displayName) {
+    checkArgument(!Strings.isNullOrEmpty(bundleId), "bundle ID must not be null or empty");
+    return
+        transform(
```
Comment (hiranya911) [inline src/main/java/com/google/firebase/projectmanagement/FirebaseProjectManagementServiceImpl.java]: Please qualify these as `ApiFutures.transform()`. In general all static references should be qualified. Guava preconditions are the only acceptable exception.

File: src/main/java/com/google/firebase/projectmanagement/FirebaseProjectManagementServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,909 @@
...
+    };
+  }
+
+  /**
+   * An {@link ApiAsyncFunction} that transforms a Long Running Operation name to an {@link IosApp}
+   * or an {@link AndroidApp} instance by repeatedly polling the server (with exponential backoff)
+   * until the App is created successfully, or until the number of poll attempts exceeds the maximum
+   * allowed.
+   */
+  private class WaitOperationFunction implements ApiAsyncFunction<String, String> {
```
Comment (hiranya911) [inline src/main/java/com/google/firebase/projectmanagement/FirebaseProjectManagementServiceImpl.java]: static

File: src/main/java/com/google/firebase/projectmanagement/FirebaseProjectManagementServiceImpl.java [hunk 14]
```diff
@@ -0,0 +1,909 @@
...
+                  projectId,
+                  settableFuture),
+              delay,
+              TimeUnit.MILLISECONDS);
+        }
+        return;
+      }
+      // The Long Running Operation API guarantees that when done == true, exactly one of 'response'
+      // or 'error' is set.
+      if (operationResponseInstance.response == null
```
Comment (hiranya911) [inline src/main/java/com/google/firebase/projectmanagement/FirebaseProjectManagementServiceImpl.java]: Null check is redundant. You've already dereferenced this once earlier.

File: src/main/java/com/google/firebase/projectmanagement/FirebaseProjectManagementServiceImpl.java [hunk 15]
```diff
@@ -0,0 +1,909 @@
...
+    @Key("shaHash")
+    private String shaHash;
+
+    @Key("certType")
+    private String certType;
+  }
+
+  /* Helper methods. */
+
+  private <T> void makeGetRequest(
```
Comment (hiranya911) [inline src/main/java/com/google/firebase/projectmanagement/FirebaseProjectManagementServiceImpl.java]: Consider introducing a new helper class. May be `HttpHelper.java`?

File: src/main/java/com/google/firebase/projectmanagement/FirebaseProjectManagementServiceImpl.java [hunk 16]
```diff
@@ -0,0 +1,909 @@
...
+    @Key("nextPageToken")
+    private String nextPageToken;
+  }
+
+  /* createAndroidApp, createIosApp */
+
+  @Override
+  public AndroidApp createAndroidApp(String projectId, String packageName, String displayName)
+      throws FirebaseProjectManagementException {
+    return blockOnFutureAndHandleExceptions(
```
Comment (hiranya911) [inline src/main/java/com/google/firebase/projectmanagement/FirebaseProjectManagementServiceImpl.java]: Ideally these blocking methods should not submit work to other threads. Can we get it to run everything on the calling thread?

```
createAndroidAppOp(projectId, packageName, displayName).call();
while (true) {
    if (isAppCreated()) {
        return result;
    }
    Thread.sleep(exponentiallyIncreasingDelay);
}
```

File: src/main/java/com/google/firebase/projectmanagement/FirebaseProjectManagementServiceImpl.java [hunk 17]
```diff
@@ -0,0 +1,909 @@
...
+      }
+      if (!operationResponseInstance.done) {
+        if (numberOfPreviousPolls + 1 >= MAXIMUM_POLLING_ATTEMPTS) {
+          settableFuture.setException(createFirebaseProjectManagementException(
+              projectId,
+              "Project ID",
+              "Unable to create App: deadline exceeded.",
+              /* cause= */ null));
+        } else {
+          long delay = (long) (
```
Comment (hiranya911) [inline src/main/java/com/google/firebase/projectmanagement/FirebaseProjectManagementServiceImpl.java]: The exponential growth of the delay should be capped at some maximum. I think the Node.js implementation of this API uses a max of 10 seconds.

File: src/main/java/com/google/firebase/projectmanagement/AndroidApp.java [hunk 18]
```diff
@@ -0,0 +1,191 @@
...
+ * An instance of this class is a reference to an Android App within a Firebase Project; it can be
+ * used to query detailed information about the App, modify the display name of the App, or download
+ * the configuration file for the App.
+ */
+public class AndroidApp {
+
+  private final AndroidAppService androidAppService;
+  private final String appId;
+
+  private AndroidApp(String appId, AndroidAppService androidAppService) {
```
[the remark this answers is not in the document]
Comment (weixifan) [reply src/main/java/com/google/firebase/projectmanagement/AndroidApp.java]: An example of a thing that can be done in a factory method that you cannot do in a constructor is use a different method name for two different factory methods. The reason this is needed is we plan to have another constructor in our upcoming proposal that is also based on a String (the fully qualified resource name of the app, rather than the App ID). This is impossible with a constructor since you cannot overload the constructor with exactly the same signature.

File: src/main/java/com/google/firebase/projectmanagement/ShaCertificate.java [hunk 19]
```diff
@@ -0,0 +1,123 @@
...
+    } else if (SHA256_PATTERN.matcher(shaHash).matches()) {
+      return ShaCertificateType.SHA_256;
+    }
+    throw new IllegalArgumentException("Invalid SHA hash, it is neither SHA-1 nor SHA-256.");
+  }
+
+  /**
+   * Returns the fully qualified resource name of this SHA certificate.
+   */
+  String getName() {
```
Comment (weixifan) [inline src/main/java/com/google/firebase/projectmanagement/ShaCertificate.java]: These should be public. Please see the API proposal document for the precise public API.

File: src/main/java/com/google/firebase/projectmanagement/FirebaseProjectManagementServiceImpl.java (hunk 15 above)
Comment (hiranya911) [reply src/main/java/com/google/firebase/projectmanagement/FirebaseProjectManagementServiceImpl.java]: No, not a project-wide helper class. Just a helper for this package/class. Basically get all the http cruft out of this class.

File: src/main/java/com/google/firebase/projectmanagement/FirebaseProjectManagementServiceImpl.java (hunk 16 above)
Comment (hiranya911) [reply src/main/java/com/google/firebase/projectmanagement/FirebaseProjectManagementServiceImpl.java]: Spawning threads is expensive (and even restricted) in some environments. The old GAE runtime is a good example, where there are strict caps on how many threads can be spawned by an app. Therefore we have designed our blocking methods to never spawn additional threads.

File: src/main/java/com/google/firebase/projectmanagement/ShaCertificateType.java [hunk 20]
```diff
@@ -19,6 +19,8 @@
  * Enum denoting types of SHA certificates currently supported by Firebase.
  */
 public enum ShaCertificateType {
+  /** Certificate generated by SHA-1 hashing algorithm. */
   SHA_1,
```
[the remark this answers is not in the document]
Comment (weixifan) [reply src/main/java/com/google/firebase/projectmanagement/ShaCertificateType.java]: According to our own official Public API documentation, we are using the underscore in our constants. https://firebase.google.com/docs/projects/api/reference/rest/v1beta1/projects.androidApps.sha#ShaCertificateType

**Comment by weixifan** [comment]
Thanks for the review. Aside from the asynchronous polling question that remains, we have another small issue: I noticed that for the `AndroidApp.deleteShaCertificate()` method, the caller must pass in a `ShaCertificate`, constructed with the `name` field set. This means we either force the caller to only be able to pass in a `ShaCertificate` returned by the `getShaCertificates()` method, or we should expose a public method allowing them to create a `ShaCertificate` with a caller-supplied `name` field. I think we should allow the user to delete Android SHA certificates without having to call `getShaCertificates` first, if they happen to know the full resource name of the certificate they want to delete prior to calling it, but this isn't that critical. Please let me know what you think.

File: src/main/java/com/google/firebase/projectmanagement/ShaCertificateType.java (hunk 20 above)
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/main/java/com/google/firebase/projectmanagement/ShaCertificateType.java]: In plain English it's normally written as SHA1 or as SHA-1 (see https://tools.ietf.org/html/rfc3174). I kind of think `SHA1` would have been better for our API, but it's not a big deal (especially since this is an enum value and not a class name).

File: src/main/java/com/google/firebase/ImplFirebaseTrampolines.java [hunk 21]
```diff
@@ -67,6 +68,10 @@ public static ThreadFactory getThreadFactory(@NonNull FirebaseApp app) {
     return app.getThreadFactory();
   }
 
+  public static ScheduledExecutorService getScheduledExecutorService(@NonNull FirebaseApp app) {
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/main/java/com/google/firebase/ImplFirebaseTrampolines.java]: I cannot seem to find any comment related to this. Did you forget to submit a comment?

[the remark this answers is not in the document]
Comment (weixifan) [reply src/main/java/com/google/firebase/ImplFirebaseTrampolines.java]: Sorry, I was referring to the comment in line 259 of `FirebaseProjectManagementServiceImpl.java`.

File: src/main/java/com/google/firebase/projectmanagement/FirebaseProjectManagementServiceImpl.java (hunk 16 above)
Comment (hiranya911) [reply src/main/java/com/google/firebase/projectmanagement/FirebaseProjectManagementServiceImpl.java]: Individual services should not spawn new threads. None of the other services (auth, fcm, iid) do, and I don't see any reason why this service should be different. Also methods like `createAndroidApp` are unlikely to be called at high volumes for performance to ever become a problem. Therefore please implement the simple/naive solution with blocking threads for now. If it ever comes to that, we can make an exception for this service and allow it to spawn threads.

Comment (weixifan) [reply src/main/java/com/google/firebase/projectmanagement/FirebaseProjectManagementServiceImpl.java]: Thanks, good point. In fact, after thinking about this some more, I think more generally the only surface through which `ThreadManager` should affect threading is the asynchronous API, rather than the blocking API, so that clients that wish to have programmatic threading control should exclusively use our asynchronous API. It is indeed unnatural to force the client to choose a particular `ThreadManager` just so that their blocking calls do not spawn threads.

For this blocking version of `createAndroidApp`, I will use your original suggestion of blocking using just `Thread.sleep(...)` (will push shortly). As for my proposed asynchronous `createAndroidAppAsync` implementation, please let me know if you have any further concerns.

Comment (hiranya911) [reply src/main/java/com/google/firebase/projectmanagement/FirebaseProjectManagementServiceImpl.java]: My recommendation would be to keep things absolutely simple. Once you have a blocking implementation based on `Thread.sleep()`, the async version of the method can just run the blocking implementation on a thread provisioned by the `ThreadManager` -- i.e. the same way the other async methods work. We can implement something more sophisticated later, if the need ever arises.

File: src/main/java/com/google/firebase/projectmanagement/FirebaseProjectManagementServiceImpl.java (hunk 4 above)
Comment (weixifan) [reply src/main/java/com/google/firebase/projectmanagement/FirebaseProjectManagementServiceImpl.java]: This was moved to `HttpHelper`. Resolving.

File: src/main/java/com/google/firebase/projectmanagement/FirebaseProjectManagementServiceImpl.java (hunk 5 above)
Comment (weixifan) [reply src/main/java/com/google/firebase/projectmanagement/FirebaseProjectManagementServiceImpl.java]: As we discussed, I have re-implemented the blocking API with `Thread.sleep()`, and instead of exposing the entire `ScheduledExecutorService` from the `FirebaseApp`, I have instead exposed three separate methods `transform`, `transformAsync`, and `schedule`.

File: src/main/java/com/google/firebase/projectmanagement/FirebaseProjectManagementServiceImpl.java (hunk 10 above)
Comment (weixifan) [reply src/main/java/com/google/firebase/projectmanagement/FirebaseProjectManagementServiceImpl.java]: Done.

File: src/main/java/com/google/firebase/ImplFirebaseTrampolines.java (hunk 21 above)
[the remark this answers is not in the document]
Comment (weixifan) [reply src/main/java/com/google/firebase/ImplFirebaseTrampolines.java]: This is resolved as per our discussion in the above comments.

File: src/main/java/com/google/firebase/projectmanagement/FirebaseProjectManagementServiceImpl.java (hunk 16 above)
Comment (weixifan) [reply src/main/java/com/google/firebase/projectmanagement/FirebaseProjectManagementServiceImpl.java]: I will consolidate this thread with the one above. Resolving this one.

**Comment by weixifan** [comment]
Thank you for the reviews. I have added the client version header as we discussed.
